### PR TITLE
Remove the null=True on CharField fields of the PackageURLMixin #36

### PR DESCRIPTION
**Pull request by tdruez**
From https://docs.djangoproject.com/en/3.0/ref/models/fields/#null

> Avoid using null on string-based fields such as CharField and TextField. If a string-based field has null=True, that means it has two possible values for “no data”: NULL, and the empty string. In most cases, it’s redundant to have two possible values for “no data;” the Django convention is to use the empty string, not NULL. 

This is an issue when adding database constraints based on those fields as '' and null values do not behave the same.

Add a `empty=None` param on the `PackageURL.to_dict` method to allow custom value like empty string '' in place of None

Signed-off-by: Thomas Druez <tdruez@nexb.com>

### CODE REVIEW

File: src/packageurl/contrib/django_models.py [hunk 1]
```diff
@@ -114,16 +108,20 @@ def package_url(self):
...
+        for field_name, value in package_url_dict.items():
             model_field = self._meta.get_field(field_name)
 
             if value and len(value) > model_field.max_length:
-                raise ValidationError(_('Value too long for field "{}".'.format(field_name)))
+                message = _('Value too long for field "{}".'.format(field_name))
+                raise ValidationError(message)
 
-            setattr(self, field_name, value or None)
+            setattr(self, field_name, value)
```
Comment (pombredanne) [inline src/packageurl/contrib/django_models.py:127]: Have you considered `setattr(self, field_name, value or '')`?

File: src/packageurl/__init__.py [hunk 2]
```diff
@@ -285,16 +285,21 @@ def __str__(self, *args, **kwargs):
...
+        If `encode` is True, then "qualifiers" are encoded as a normalized
+        string. Otherwise, qualifiers is a mapping.
+        You can provide a value for `empty` to be used in place of default None.
         """
         data = self._asdict()
         if encode:
             data['qualifiers'] = normalize_qualifiers(self.qualifiers,
                                                       encode=encode)
+
+        if empty is not None:
```
Comment (pombredanne) [inline src/packageurl/__init__.py:300]: Why using only there?

File: src/packageurl/__init__.py [hunk 3]
```diff
@@ -285,16 +285,21 @@ def __str__(self, *args, **kwargs):
...
+            data = OrderedDict((k, v or empty) for k, v in data.items())
```
Comment (pombredanne) [inline src/packageurl/__init__.py:301]: Since data is already an OrderedDict... you may want to modify it in place but this is minor:
```
for k, v in data.items():
    data[k] = v or empty
```

File: src/packageurl/contrib/django_models.py (hunk 1 above)
Comment (tdruez) [reply src/packageurl/contrib/django_models.py:127]: Yes I did. It was my initial approach, but the empty value is now controlled in the above `to_dict` which makes it unneeded in the `setattr`.

File: src/packageurl/__init__.py (hunk 2 above)
Comment (tdruez) [reply src/packageurl/__init__.py:300]: Could you rephrase your question?

File: src/packageurl/__init__.py (hunk 3 above)
Comment (tdruez) [reply src/packageurl/__init__.py]: Changed in https://github.com/package-url/packageurl-python/pull/40/commits/f2e962fe97801cfe2ecfdc6f6cec9850aced93df

**Comment by sbs2001** [comment]
@pombredanne  This makes total sense and is definitely needed. It will prevent https://github.com/nexB/vulnerablecode/issues/206#issuecomment-649293366 from happening.

Thanks @tdruez

**Comment by tdruez** [comment]
@sbs2001 Thanks for your input 👍

File: src/packageurl/__init__.py (hunk 2 above)
Comment (pombredanne) [reply src/packageurl/__init__.py:300]: Sorry for the poor wording ... I mean why would you only translate the empty values to `empty` if `empty` is not None?
Why not doing it all the time and therefore removing the test `if empty is not None:`?
